crypten/mpc/primitives/resharing.py

```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import crypten
+import crypten.communicator as comm
+import torch
+
+from crypten.cuda import CUDALongTensor
+from crypten.common.util import torch_stack
+from crypten.common.rng import generate_kbit_random_tensor, generate_random_ring_element
+from crypten.encoder import FixedPointEncoder
+
+import time
+
+def replicate_shares(x_share):
+    """
+    Replicate additively secret-shared tensor x。
+    Party i sends x_i to party i+1, and receives
+    x_{i-1} from party i-1.
+    Each party holds 2-out-of-3 secret sharing of tensor
+    x after the protocol. 
+    """
+    rank = comm.get().get_rank()
+    world_size = comm.get().get_world_size()
+    prev_rank = (rank - 1) % world_size
+    next_rank = (rank + 1) % world_size
+
+    x_share = x_share.contiguous()
+    x_rep = torch.zeros_like(x_share.data)
+
+    req1 = comm.get().isend(x_share.data, dst=next_rank)
+    req2 = comm.get().irecv(x_rep.data, src=prev_rank)
+
+    req1.wait()
+    req2.wait()
+
+    return x_rep
+
+
+def __replicated_secret_sharing_protocol(op, x, y, *args, **kwargs):
+    """
+    Implement Replicated Secret Sharing (RSS) protocol for 3PC computation.
+    Each party holds 2-out-of-3 secret sharing of tnesor x and y. 
+    See Section 3.2.1 of "ABY3 : A Mixed Protocol Framework for Machine Learning"
+    for more detail.
+    """
+    assert op in {
+        "mul",
+        "matmul",
+        "conv1d",
+        "conv2d",
+        "conv_transpose1d",
+        "conv_transpose2d",
+    }
+    from .arithmetic import ArithmeticSharedTensor
+    from .binary import BinarySharedTensor
+
+    x1, x2 = x.share, replicate_shares(x.share)
+    y1, y2 = y.share, replicate_shares(y.share)
```

Comment:
We may want the ability to cache the replicated shares since they can be fixed per variable. This can save communication when there are computations using any repeated variables.

This can go in a separate diff and I can handle it if it makes things easier.

Reply:
Definitely. I currently can't think of an elegant API of doing so. Once we land this diff, we can discuss how we could enable caching.